src/utilities/timeline.ts

```diff
@@ -906,97 +942,213 @@ export function spanInView(span: Span, viewTimeRange: TimeRange) {
+      Array.from(allKeys)
+        .sort()
+        .forEach(type => {
+          const externalEventsGroup = groupedExternalEvents[type];
+          const id = type;
+          const expanded = getNodeExpanded(id, discreteTreeExpansionMap);
+          const label = type;
+          const children: DiscreteTreeNode['children'] = [];
+          const items: DiscreteTreeNode['items'] = [];
+          if (externalEventsGroup) {
+            externalEventsGroup.forEach(externalEvent => {
+              items.push({ externalEvent });
+              children.push({
+                activity_type: undefined, // ignored.
+                children: [],
+                expanded: expanded,
+                id: `${externalEvent.pkey.key}`,
+                isLeaf: true,
+                items: [{ externalEvent }],
+                label: externalEvent.pkey.key,
+                type: 'ExternalEvent',
+              });
+            });
+          }
+          externalEventNodes.push({
+            activity_type: undefined, // ignored.
+            children: paginateNodes(children, 'externalEvent', id, discreteTreeExpansionMap),
+            expanded: expanded,
+            id,
+            isLeaf: false,
+            items: items,
+            label,
+            type: 'ExternalEvent',
+          });
+        });
+    }
+  }
+
+  // if both are present, cluster them
+  if (hasActivityLayer && activityNodes.length && hasExternalEventsLayer && externalEventNodes.length) {
+    const activityAggNode: DiscreteTreeNode = {
+      activity_type: undefined,
+      children: activityNodes,
+      expanded: getNodeExpanded('!!activity-agg', discreteTreeExpansionMap),
+      id: '!!activity-agg',
+      isLeaf: false,
+      items: flattenItems(activityNodes).filter(
+        (obj, index, self) => index === self.findIndex(o => o.directive?.id === obj.directive?.id),
+      ),
+      label: 'Activities',
+      type: 'Activity', // RowHeaderDiscreteTree does not seem to require any special treatment; so no special category for this top node
+    };
+    const externalEventAggNode: DiscreteTreeNode = {
+      activity_type: undefined,
+      children: externalEventNodes,
+      expanded: getNodeExpanded('!!ex-ev-agg', discreteTreeExpansionMap),
+      id: '!!ex-ev-agg',
+      isLeaf: false,
+      items: flattenItems(externalEventNodes).filter(
+        (obj, index, self) => index === self.findIndex(o => o.externalEvent?.pkey === obj.externalEvent?.pkey),
+      ),
+      label: 'External Events',
+      type: 'ExternalEvent', // RowHeaderDiscreteTree does not seem to require any special treatment; so no special category for this top node
+    };
+    return [activityAggNode, externalEventAggNode];
+  } else if (activityNodes.length) {
+    return activityNodes;
+  } else {
+    return externalEventNodes;
+  }
+}
+
+export function flattenItems(nodes: DiscreteTreeNode[]): DiscreteTreeNodeItem[] {
```

Comment:
Can you leave a comment about what this function and the helper do?

Reply:
And potentially unit tests?

Reply:
This function has been removed in favor of a faster map; see [replace flattenItems with faster map](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/a0e7c826d9782eb278bf4a79c11284aecd013f0d). This should also address the comment made on the PR about efficiency concerns regarding `flattenItems`.